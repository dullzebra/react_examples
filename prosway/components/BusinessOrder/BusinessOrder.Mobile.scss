:local {
  @media
  (max-width: $widthMobileScreenPortrait) and (orientation: portrait),
  (max-width: $widthMobileScreenLandscape) and (orientation: landscape) {

    .BusinessOrder {
      width: auto;
      padding: $step;
    }

    .BusinessOrderMenu {
      margin-left: -$step;
    }
  }
}
