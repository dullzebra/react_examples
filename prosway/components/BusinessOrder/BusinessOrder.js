import React, { Component } from 'react';
import { connect } from 'react-redux';
import qs from 'qs';

import s from './BusinessOrder.scss';
import Tabs from '_components/Tabs/Tabs';
import OrdersList from './OrdersList/OrdersList';
import Button from '_uicomponents/Button/Button';
import { fetchUserOrders, loadMoreUserOrders } from '_actions';
import { params } from '_config/config';
import BusinessOrderCalendar from './BusinessOrderCalendar/BusinessOrderCalendar';

export class BusinessOrder extends Component {
  state = {
    activeTab: 0,
    loadingMore: false,
    nextUrl: null,
  };

  componentDidMount() {
    this.props.fetchUserOrders(params.businessOrder)
      .then(data => {
        this.setState({ nextUrl: data.next });
      });
  }

  loadMore = () => {
    const { nextUrl } = this.state;
    const loadMoreParams = qs.parse(nextUrl.split('/?')[1]);

    this.setState({ loadingMore: true });
    this.props.loadMoreUserOrders({ ...params.businessOrder, ...loadMoreParams })
      .then(data => {
        this.setState({ loadingMore: false, nextUrl: data.next });
      });
  };

  onTabClick = (i) => {
    this.setState({ activeTab: i });
  };

  renderOrdersList = () => {
    const { orders } = this.props;
    const { loadingMore, nextUrl } = this.state;

    return (
      <React.Fragment>
        <OrdersList items={orders}/>
        {!!nextUrl && <div className={s.BusinessOrderButtons}>
          {loadingMore
          && <Button
            bg={'transparent'}
            type='button'
          >
            Подождите
          </Button>
          }
          {orders && !loadingMore
          && <Button
            bg={'transparent'}
            onClick={this.loadMore}
            type='button'
          >
            Показать ещё
          </Button>
          }
        </div>}
      </React.Fragment>
    );
  };

  render() {
    const { activeTab } = this.state;

    return (
      <div className={s.BusinessOrder}>
        <h1 className={s.BusinessOrderTitle}>
          Мои заказы
        </h1>
        <div className={s.BusinessOrderMenu}>
          <Tabs
            activeTab={activeTab}
            tabs={[
              { label: 'Заявки' },
              { label: 'Календарь' },
            ]}
            onTabClick={this.onTabClick}
            customCssClass='Tabs_themeTall'
          />
        </div>
        {activeTab === 0 && this.renderOrdersList()}

        {activeTab === 1 && <BusinessOrderCalendar timeStart={9} timeEnd={21}/>}
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  const { user: { orders } } = state;

  return {
    orders,
  };
};

export default connect(mapStateToProps, { fetchUserOrders, loadMoreUserOrders })(BusinessOrder);
