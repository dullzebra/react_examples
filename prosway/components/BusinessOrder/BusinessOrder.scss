@import '~_styles/variables.scss';

:local {
  .BusinessOrder {
    position: relative;
    width: $widthLayoutBig;
    margin: 0 auto ($step * 3);
    padding: $step $stepBig;
    background-color: $colorWhite;
    box-shadow: 0 12px 24px 0 rgba(0, 0, 0, 0.06);
  }

  .BusinessOrderTitle {
    margin-bottom: $step;
    font-size: 24px;
  }

  .BusinessOrderMenu {
    margin-bottom: $step;
  }

  .BusinessOrderButtons {
    display: flex;
    justify-content: center;
  }
}

@import './BusinessOrder.Mobile.scss';
