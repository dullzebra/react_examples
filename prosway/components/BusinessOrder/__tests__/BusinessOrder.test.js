import React from 'react';
import renderer from 'react-test-renderer';
import { combineReducers } from 'redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import userReducer from '../../../../reducers/userReducer';
import { fetchUserOrders } from '_actions/';
import { FETCH_USER_ORDERS } from '_actions/types';
import {BusinessOrder} from '../BusinessOrder';
import orderResponse from '../__mock__/orderResponse';
import orderStore from '../__mock__/orderStore';

jest.mock('../OrdersList/OrdersList', () => () => 'OrdersList');

const mockFn = jest.fn(() => Promise.resolve({results:[], next: 'next'}));

const expectedProps = {
  fetchUserOrders: mockFn,
  orders: [],
};

const setupComponent = () => {
  return renderer.create(
    <BusinessOrder {...expectedProps} />
  );
};

const axiosInstance = axios.create();
const middleware = [thunk.withExtraArgument(axiosInstance)];
const mockStore = configureMockStore(middleware);
const reducerMock = combineReducers({user: userReducer});

describe('<BusinessOrder />', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render w/o crashing', () => {
    const component = setupComponent();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call fetchUserOrders', () => {
    expect(mockFn).toHaveBeenCalled();
  });

  it('should render <Показать ещё> button after fetchUserOrders call', () => {
    const component = setupComponent();
    mockFn().then(() => {
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('BusinessOrder: async actions test', () => {
  let mock;
  const initState = {
    user: {
      orders: [],
    },
  };

  beforeEach(() => {
    mock = new MockAdapter(axiosInstance);
    mock.onGet('order/').reply(200, orderResponse);
  });

  afterEach(() => {
    mock.restore();
  });

  it('should create action FETCH_USER_ORDERS and put response items to store ', done => {
    const expectedAction = {
      type: FETCH_USER_ORDERS,
      payload: orderResponse,
    };

    let state = initState;
    const store = mockStore(() => state);
    store.dispatch(fetchUserOrders({}));

    setTimeout(() => {
      const actions = store.getActions();
      //console.log(actions);
      expect(actions).toContainEqual(expectedAction);

      state = reducerMock(state, expectedAction);
      //console.log(state);
      expect(state).toEqual({
        ...initState,
        user: {
          orders: orderStore,
        },
      });
      done();
    }, 1000);
  });
});
