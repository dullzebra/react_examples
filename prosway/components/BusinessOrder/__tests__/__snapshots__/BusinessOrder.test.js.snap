// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BusinessOrder /> should render <Показать ещё> button after fetchUserOrders call 1`] = `
<div
  className="BusinessOrder"
>
  <h1
    className="BusinessOrderTitle"
  >
    Мои заказы
  </h1>
  <div
    className="BusinessOrderMenu"
  >
    <div
      className="Tabs Tabs_themeTall"
    >
      <div
        className="Bar"
      >
        <div
          className="Tab active"
          onClick={[Function]}
        >
          Заявки
        </div>
        <div
          className="Tab"
          onClick={[Function]}
        >
          Календарь
        </div>
        <div
          className="Indicator"
          style={
            Object {
              "transform": "translateX(0px)",
              "width": 0,
            }
          }
        />
      </div>
    </div>
  </div>
  OrdersList
  <div
    className="BusinessOrderButtons"
  >
    <button
      className="ButtonTransparent"
      disabled={undefined}
      onClick={[Function]}
      type="button"
    >
      Показать ещё
    </button>
  </div>
</div>
`;

exports[`<BusinessOrder /> should render w/o crashing 1`] = `
<div
  className="BusinessOrder"
>
  <h1
    className="BusinessOrderTitle"
  >
    Мои заказы
  </h1>
  <div
    className="BusinessOrderMenu"
  >
    <div
      className="Tabs Tabs_themeTall"
    >
      <div
        className="Bar"
      >
        <div
          className="Tab active"
          onClick={[Function]}
        >
          Заявки
        </div>
        <div
          className="Tab"
          onClick={[Function]}
        >
          Календарь
        </div>
        <div
          className="Indicator"
          style={
            Object {
              "transform": "translateX(0px)",
              "width": 0,
            }
          }
        />
      </div>
    </div>
  </div>
  OrdersList
</div>
`;
